fmf/utils.py

```diff
@@ -735,7 +739,18 @@ def fetch_repo(url, ref=None, destination=None, env=None):
             if age >= CACHE_EXPIRATION:
                 run(['git', 'fetch'], cwd=destination)
             # Checkout branch
-            run(['git', 'checkout', '-f', ref], cwd=destination, env=env)
+            try:
+                run(['git', 'checkout', '-f', ref], cwd=destination, env=env)
+            except subprocess.CalledProcessError as error:
+                # Another way to check for shallow clone is to call
+                # `git rev-parse --is-shallow-repository` but it prints true/false
+                if os.path.isfile(os.path.join(destination, '.git', 'shallow')):
+                    # Fetch whole history and try again
+                    run(['git', 'fetch', '--unshallow'], cwd=destination)
```

Comment:
```suggestion
                    run(['git', 'fetch', '--unshallow', 'origin', '+refs/heads/*:refs/heads/*'], cwd=destination)
```

Seems we have to explicitly ask to fetch remote branches as well otherwise the second checkout fails:

    fmf.utils.fetch_repo("https://github.com/teemtee/tmt/", destination="/tmp/fetched")
    fmf.utils.fetch_repo("https://github.com/teemtee/tmt/", destination="/tmp/fetched", ref="fedora")
    error: pathspec 'fedora' did not match any file(s) known to git

Perhaps add a simple test to cover this?

Reply:
I definitely will add tests (also to check that shallow clone happened). Thanks for suggestion - I've tried with commit hash and that works without suggested change.

Reply:
It was not enough, I had to follow https://www.delenamalan.co.za/2019/fetch-all-branches.html